File: 0768-partition-labels/0768-partition-labels.erl
-spec partition_labels(S :: unicode:unicode_binary()) -> [integer()].
partition_labels(<<S/binary>>) -> partition_labels(binary_to_list(S));
partition_labels(S) ->
    PosMap = element(1, lists:foldl(
        fun(X, {Map, Idx}) ->
            {maps:update_with(X, fun(_) -> Idx end, Idx, Map), Idx + 1}
        end, {maps:new(), 0}, S
    )),
    element(1, lists:foldl(
        fun(X, {Result, Prev, Max, Idx}) ->
            NewMax = max(Max, maps:get(X, PosMap)),
            if
                Idx == NewMax ->
                    {Result ++ [NewMax - Prev], NewMax, NewMax, Idx + 1};
                true ->
                    {Result, Prev, NewMax, Idx + 1}
            end
        end, {[], -1, 0, 0}, S
    )).

File: 0768-partition-labels/0768-partition-labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        const int n = s.length();
        int pos[26] = {0,};

        for(int i = 0; i < n; ++i){
            pos[s[i] - 'a'] = i;
        }

        vector<int> ans{};
        int prev = -1, max = 0;
        for(int i = 0; i < n; ++i){
            max = ::max(max, pos[s[i] - 'a']);
            if(max == i){
                ans.push_back(max - prev);
                prev = max;
            }
        }

        return ans;
    }
};